README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# IPO2

<!-- badges: start -->
<!-- badges: end -->

Isotopologue Parameter Optimization (IPO) is a commonly used tool to
optimize parameters for XCMS analysis of metabolomics data. The progress
in XCMS development has made it more difficult to utilize IPO, and so
IPO2 is an effort to more seamlessly integrate this method of parameter
optimization with the latest version of XCMS. While this package adopts
a similar approach of using isotope identification, scoring, and
design-of-experiments, the underlying code to accomplish these tasks is
different and the results are likely not exactly the same.

Please find a copy of the original IPO manuscript in the source of this
package.

``` r
citation("IPO")
#> 
#> To cite the IPO package in publications use:
#> 
#>   Libiseller, Dvorzak, Kleb, Gander, Eisenberg, Madeo, Neumann,
#>   Trausinger, Sinner, Pieber, Magnes IPO: A Tool for automated
#>   Optimization of XCMS Parameters BMC Bioinformatics (accepted)
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     author = {Gunnar Libiseller and Michaela Dvorzak and Ulrike Kleb and Edgar Gander and Tobias Eisenberg and Frank Madeo and Steffen Neumann and Gert Trausinger and Frank Sinner and Thomas Pieber and Christoph Magnes},
#>     title = {IPO: a tool for automated optimization of XCMS parameters},
#>     journal = {BMC Bioinformatics},
#>     year = {(2015)},
#>     volume = {16},
#>     pages = {118},
#>   }
```

## Installation

You can install the original release version of IPO from
[Bioconductor](https://www.bioconductor.org/packages/release/bioc/html/IPO.html).

``` r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("IPO")
```

And the development version is available from
[GitHub](https://github.com/rietho/IPO).

``` r
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
    
devtools::install_github("rietho/IPO")
```

You can install the development version of IPO2 from
[GitHub](https://github.com/rietho/IPO2) with:

``` r
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
    
devtools::install_github("wmoldham/IPO2")
```

## Usage


NEWS.md
# IPO2 0.0.4

* Mandatory output in working directly by default.
* Save history as `.csv` file. 

# IPO2 0.0.3

* Add history to log file output in case of failed run.
* Split sink output to both log file and console.
* Increase grouping parameter bandwidth minimum to 1 (from 0.01) to increase processing speed.

# IPO2 0.0.2

* Reformat renaming of log files.
* Handle peak scoring when no isotope peaks identified.
* Combine all steps in optimization to return score only.

# IPO2 0.0.1

* Working `optimize_centwave`.
* Working `optimize_align_group`.


tests/testthat.R
library(testthat)
library(IPO2)

test_check("IPO2")


tests/testthat/test-utils.R
test_that("%nin% works", {
  expect_equal((c(1, 2, 3) %nin% c(3, 4, 5)), c(TRUE, TRUE, FALSE))
})


tests/testthat/test-optimize_centwave.R
test_that("check_centwave_params works", {
  expect_error(check_centwave_params(list(foo = "bar")), "These parameters are not used by centWave: ")
  expect_error(check_centwave_params(list(ppm = c(1, 2, 3))), "Too many parameters specified for: ")
  expect_error(check_centwave_params(list(ppm = 1)), "No parameters specified for optimization")
  expect_error(check_centwave_params(list(ppm = c(-1, 1))), "must be greater than 0")
  expect_error(check_centwave_params(list(snthresh = c(-1, 1))), "must not be negative")
})


R/utils.R
"%nin%" <- function(x, table) {
  match(x, table, nomatch = 0L) == 0L
}

prepare_out_dir <- function(out_dir = NULL, fun_name) {

  if (is.null(out_dir)) out_dir <- getwd()
  if(!dir.exists(out_dir)) {
    cat("Creating", out_dir, "\n")
    dir.create(out_dir, recursive = TRUE)
  }

  dir <- paste0(out_dir, "/", fun_name, format(Sys.time(), "_%Y-%m-%d-%H-%M-%S"))

  dir.create(dir)
  assign("dir", paste0(dir, "/"), envir = parent.frame())
  assign("log_file", paste0(dir, "/log_file.txt"), envir = parent.frame())

}

parse_parameters <- function(
  parameter_list,
  method = c("centwave", "align_group")
) {

  quant <- qual <- lists <- NULL

  default_params <- switch(
    method,
    "centwave" = {
      param_types_centwave()
      get_centwave_defaults()
    },
    "align_group" = {
      param_types_align_group()
      c(get_obiwarp_defaults(), get_density_defaults())
    }
  )

  default_params[names(parameter_list)] <- parameter_list

  params <- list()
  params$to_optimize <- default_params

  params <- list()
  params$to_optimize <- Filter(function(x) length(x) == 2, default_params)
  params$constant <- Filter(function(x) length(x) != 2, default_params[c(quant, qual)])
  params$lists <- default_params[lists]

  params
}


generate_ccd <- function(parameters_to_optimize) {

  lower_bound <- sapply(parameters_to_optimize, "[[", 1)
  upper_bound <- sapply(parameters_to_optimize, "[[", 2)
  center <- (upper_bound - lower_bound) / 2

  x <- paste0(
    "x",
    1:length(parameters_to_optimize),
    " ~ (",
    c(names(parameters_to_optimize)),
    " - ",
    (lower_bound + center),
    ") / ",
    center
  )

  formulas <- lapply(x, stats::as.formula)

  rsm::ccd(
    length(parameters_to_optimize),  # number of variables
    n0 = 1,  # number of center points
    alpha = "face",  # position of the ‘star’ points
    randomize = FALSE,
    inscribed = TRUE,  # axis points are at +/- 1 and the cube points are at interior positions
    coding = formulas  # list of coding formulas for the design variables
  )

}


design_experiments <- function(parameters) {

  run.order <- std.order <- Block <- NULL

  if(length(parameters$to_optimize) > 1) {
    design <- generate_ccd(parameters$to_optimize)
    design <- rsm::decode.data(design)
    design <- subset(design, select = -c(run.order, std.order, Block))
  } else {
    design <- seq(min(parameters$to_optimize[[1]]),
                  max(parameters$to_optimize[[1]]),
                  length.out = 9)
    design <- data.frame(design)
    colnames(design) <- names(parameters$to_optimize[1])
  }

  data.table(design)

}


create_model <- function(design, score) {

  params <- paste0(colnames(design), collapse = ", ")
  design <- cbind(design, score = score$score)

  if(ncol(design) > 1) {
    formula <- stats::as.formula(paste0("score ~ SO(", params, ")"))
    model <- rsm::rsm(formula, data = design)
  } else {
    formula <- stats::as.formula(paste("score ~", params, "+", params, "^ 2"))
    model <- stats::lm(formula, data = design)
  }

  model

}


get_maximum <- function(design, model) {

  number_params <- ncol(design)
  steps <- ceiling(1E6 ^ (1/number_params))

  param_values <- list()
  for (nm in names(design)) {
    minimum <- min(design[[nm]])
    maximum <- max(design[[nm]])
    param_values[[nm]] <- seq(minimum, maximum, length.out = steps)
  }

  search_grid <- do.call(CJ, param_values)
  max_value <- stats::predict(model, search_grid)
  unlist(search_grid[max_value == max(max_value), ][1, ])

}


plot_contours <- function(design, model, maximum, plot_name) {

  # make formula
  form <- stats::as.formula(
    paste("~ ", paste(colnames(design), collapse = " + "))
  )

  # set plotting area
  number_params <- 1:8
  number_pairs <- number_params * (number_params - 1) / 2
  number_rows <- c(1, 1, 1, 2, 2, 3, 3, 4)
  number_columns <- number_pairs / number_rows

  params <- ncol(design)
  cols <- number_columns[which(number_params == params)]
  rows <- number_rows[which(number_params == params)]

  grDevices::png(
    plot_name,
    width = cols * 4,
    height = rows * 4,
    units = "in",
    res = 300
  )

  graphics::par(mfrow = c(rows, cols))

  # plot
  graphics::contour(model, form = form, image = TRUE, at = maximum)
  grDevices::dev.off()

}


pick_parameters <- function(parameters, maximum) {

  width <- purrr::map_dbl(parameters, diff)
  delta <- purrr::map2(parameters, as.list(maximum), ~abs(.x - .y))
  params <- list()

  boundaries <- list(
    ppm = c(0, Inf, 2),
    min_peakwidth = c(0, Inf, 2),
    max_peakwidth = c(0, Inf, 2),
    snthresh = c(0, Inf, 0),
    noise = c(0, Inf, 0),
    prefilter_k = c(0, Inf, 0),
    prefilter_int = c(0, Inf, 0),
    mzdiff = c(-Inf, Inf, 4),
    binSize_O = c(0, Inf, 3),
    response = c(1, 100, 0),
    gapInit = c(0, Inf, 2),
    gapExtend = c(0, Inf, 2),
    factorDiag = c(-Inf, Inf, 0),
    factorGap = c(-Inf, Inf, 0),
    initPenalty = c(0, Inf, 2),
    bw = c(1, Inf, 2),
    minFraction = c(0, 1, 2),
    minSamples = c(0, Inf, 0),
    binSize_D = c(0.001, Inf, 3),
    maxFeatures = c(0, Inf, 2)
  )

  for (nm in names(parameters)) {

    if (sum(delta[[nm]] == 0) > 0) {
      width[[nm]] <- width[[nm]] * 1.4
    } else {
      width[[nm]] <- width[[nm]] * 0.8
    }

    upper <- maximum[[nm]] + 0.5 * width[[nm]]
    lower <- maximum[[nm]] - 0.5 * width[[nm]]

    lower <- max(boundaries[[nm]][[1]], lower)
    upper <- min(boundaries[[nm]][[2]], upper)

    lower <- round(lower, digits = boundaries[[nm]][[3]])
    upper <- round(upper, digits = boundaries[[nm]][[3]])

    params[[nm]] <- unique(c(lower, upper))  # stop optimizing if converge

  }

  # prevent overlapping peakwidth optimization
  if (sum(
    is.null(params[["min_peakwidth"]]), is.null(params[["max_peakwidth"]])) == 0
  ){
    if (max(params[["min_peakwidth"]] > min(params[["max_peakwidth"]]))) {
      params[["max_peakwidth"]][[1]] <- params[["min_peakwidth"]][[2]]
    }
  }

  params

}


retry_parallel <- function(fun) {
  redo <- TRUE
  trial <- 1
  redo_list <- list()
  while(redo & trial <= 5) {
    out <-
      BiocParallel::bptry(
        eval(fun, envir = rlang::env(rlang::caller_env(), redo_list = redo_list))
      )
    errs <- sum(BiocParallel::bpok(out) == FALSE)
    redo <- errs > 0
    if (errs > 0) {
      ids <- which(BiocParallel::bpok(out) == FALSE)
    } else {
      ids <- vector()
    }
    cat(
      "     Trial:", trial,
      "     Errors:", sprintf("%3d", errs),
      "     IDs:", ids,
      "\n"
    )
    if (redo) {
      redo_list <- out
      trial <- trial + 1
    }
  }
  out
}


R/score_peaks.R
score_peaks <- function(xcmsnexp) {

  total <- peak_tables <- isotope_list <- indeterminate <- isotopes <- score <- NULL

  peak_df <- data.table(xcms::chromPeaks(xcmsnexp), keep.rownames = TRUE)

  if (nrow(peak_df) == 0) {
    return(
      data.table(
        total = 0,
        indeterminate = 0,
        isotopes = 0,
        score = 0
      )
    )
  }

  a <-
    peak_df[

      # summarize peak tables by sample
      , list(peak_tables = list(.SD)), by = sample
    ][

      # count total number of peaks
      , total := purrr::map_dbl(peak_tables, nrow)
    ][

      # identify isotope peaks
      , isotope_list := purrr::map(peak_tables, find_isotopes)
    ][

      # identify low intensity peaks with likely undetectable isotopes
      , indeterminate := purrr::map2_dbl(peak_tables, isotope_list, find_indeterminate_peaks)
    ][

      # count isotopes in list
      , isotopes := purrr::map_dbl(isotope_list, length)

    ]

  a[
    , lapply(.SD, sum), .SDcols = c("total", "indeterminate", "isotopes")
  ][
    , score := isotopes ^ 2 / (total - indeterminate - isotopes)
  ]
}


find_isotopes <- function(peak_table) {

  id <- NULL

  peak_table <- peak_table[order(peak_table$mz)]

  pair <- list()
  counter <- 0

  C13_mass_defect <- 1.003355
  ppm <- 5

  for (row in 1:nrow(peak_table)) {

    query <- peak_table[row]

    mz <- peak_table[[row, "mz"]]

    mz_low <- mz + C13_mass_defect - mz * ppm / 1E6
    mz_high <- mz + C13_mass_defect + mz * ppm / 1E6

    matches <- peak_table[mz >= mz_low & mz <= mz_high][, id := .I]

    if (nrow(matches) > 0) {
      counter <- counter + 1
      query_2 <- query[rep(1, nrow(matches))][, id := .I]
      pair[[counter]] <- merge(
        query_2,
        matches,
        suffixes = c("_1", "_2"),
        by = "id"
      )
    }

  }

  isotope_pairs <- data.table::rbindlist(pair)

  if (nrow(isotope_pairs) == 0) return(list())

  # identify overlapping retention time windows
  rt_overlap <-
    abs(isotope_pairs$rt_1 - isotope_pairs$rt_2) / isotope_pairs$rt_1 < 0.005

  # confirm expected peak intensity ratios
  intb_ratio <- isotope_pairs$intb_2 / isotope_pairs$intb_1
  carbon <- maximize_carbons(isotope_pairs$mz_1)
  predicted_ratio <- predict_isotope_ratio(carbon)
  expected_ratio <-
    intb_ratio >= 0.75 * predict_isotope_ratio(1) &
    intb_ratio <= 1.25 * predicted_ratio

  # list unique peak ids
  unique(
    unlist(
      isotope_pairs[rt_overlap & expected_ratio, c("rn_1", "rn_2")]
    ),
    use.names = FALSE
  )
}


maximize_carbons <- function(mass) {
  C <- 12.0
  H  <- 1.0078250170
  CH3 <- C + H * 3
  CH2 <- C + H * 2
  floor((mass - 2 * CH3) / CH2) + 2
}


predict_isotope_ratio <- function(carbons) {
  C13_abundance <- 0.0107
  stats::dbinom(1, carbons, C13_abundance) / stats::dbinom(0, carbons, C13_abundance)
}


find_indeterminate_peaks <- function(peak_tbl, isotope_list) {

  intb <- NULL

  no_isotopes <- peak_tbl[peak_tbl$rn %nin% isotope_list, ]

  three_percent <- round(nrow(no_isotopes) * 0.03)

  threshold <- no_isotopes[order(intb)][1:three_percent, mean(intb)]

  carbon <- maximize_carbons(no_isotopes$mz)
  predicted_ratio <- predict_isotope_ratio(carbon)
  predicted_intb <- no_isotopes$intb * predicted_ratio
  nrow(no_isotopes[predicted_intb < threshold, ])

}


renv/activate.R

local({

  # the requested version of renv
  version <- "0.12.3-16"

  # the project directory
  project <- getwd()

  # avoid recursion
  if (!is.na(Sys.getenv("RENV_R_INITIALIZING", unset = NA)))
    return(invisible(TRUE))

  # signal that we're loading renv during R startup
  Sys.setenv("RENV_R_INITIALIZING" = "true")
  on.exit(Sys.unsetenv("RENV_R_INITIALIZING"), add = TRUE)

  # signal that we've consented to use renv
  options(renv.consent = TRUE)

  # load the 'utils' package eagerly -- this ensures that renv shims, which
  # mask 'utils' packages, will come first on the search path
  library(utils, lib.loc = .Library)

  # check to see if renv has already been loaded
  if ("renv" %in% loadedNamespaces()) {

    # if renv has already been loaded, and it's the requested version of renv,
    # nothing to do
    spec <- .getNamespaceInfo(.getNamespace("renv"), "spec")
    if (identical(spec[["version"]], version))
      return(invisible(TRUE))

    # otherwise, unload and attempt to load the correct version of renv
    unloadNamespace("renv")

  }

  # load bootstrap tools   
  bootstrap <- function(version, library) {
  
    # read repos (respecting override if set)
    repos <- Sys.getenv("RENV_CONFIG_REPOS_OVERRIDE", unset = NA)
    if (is.na(repos))
      repos <- getOption("repos")
  
    # fix up repos
    on.exit(options(repos = repos), add = TRUE)
    repos[repos == "@CRAN@"] <- "https://cloud.r-project.org"
    options(repos = repos)
  
    # attempt to download renv
    tarball <- tryCatch(renv_bootstrap_download(version), error = identity)
    if (inherits(tarball, "error"))
      stop("failed to download renv ", version)
  
    # now attempt to install
    status <- tryCatch(renv_bootstrap_install(version, tarball, library), error = identity)
    if (inherits(status, "error"))
      stop("failed to install renv ", version)
  
  }
  
  renv_bootstrap_download_impl <- function(url, destfile) {
  
    mode <- "wb"
  
    # https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17715
    fixup <-
      Sys.info()[["sysname"]] == "Windows" &&
      substring(url, 1L, 5L) == "file:"
  
    if (fixup)
      mode <- "w+b"
  
    download.file(
      url      = url,
      destfile = destfile,
      mode     = mode,
      quiet    = TRUE
    )
  
  }
  
  renv_bootstrap_download <- function(version) {
  
    # if the renv version number has 4 components, assume it must
    # be retrieved via github
    nv <- numeric_version(version)
    components <- unclass(nv)[[1]]
  
    methods <- if (length(components) == 4L) {
      list(renv_bootstrap_download_github)
    } else {
      list(
        renv_bootstrap_download_cran_latest,
        renv_bootstrap_download_cran_archive
      )
    }
  
    for (method in methods) {
      path <- tryCatch(method(version), error = identity)
      if (is.character(path) && file.exists(path))
        return(path)
    }
  
    stop("failed to download renv ", version)
  
  }
  
  renv_bootstrap_download_cran_latest <- function(version) {
  
    repos <- renv_bootstrap_download_cran_latest_find(version)
  
    message("* Downloading renv ", version, " from CRAN ... ", appendLF = FALSE)
  
    info <- tryCatch(
      download.packages("renv", repos = repos, destdir = tempdir(), quiet = TRUE),
      condition = identity
    )
  
    if (inherits(info, "condition")) {
      message("FAILED")
      return(FALSE)
    }
  
    message("OK")
    info[1, 2]
  
  }
  
  renv_bootstrap_download_cran_latest_find <- function(version) {
  
    # check for renv on CRAN matching this version
    all <- unique(c(
      getOption("repos"),
      getOption("renv.bootstrap.repos", default = "https://cloud.r-project.org")
    ))
  
    for (repos in all) {
  
      db <- tryCatch(
        as.data.frame(available.packages(repos = repos), stringsAsFactors = FALSE),
        error = identity
      )
  
      if (inherits(db, "error"))
        next
  
      entry <- db[db$Package %in% "renv" & db$Version %in% version, ]
      if (nrow(entry) == 0)
        next
  
      return(repos)
  
    }
  
    fmt <- "renv %s is not available from your declared package repositories"
    stop(sprintf(fmt, version))
  
  }
  
  renv_bootstrap_download_cran_archive <- function(version) {
  
    name <- sprintf("renv_%s.tar.gz", version)
    repos <- getOption("repos")
    urls <- file.path(repos, "src/contrib/Archive/renv", name)
    destfile <- file.path(tempdir(), name)
  
    message("* Downloading renv ", version, " from CRAN archive ... ", appendLF = FALSE)
  
    for (url in urls) {
  
      status <- tryCatch(
        renv_bootstrap_download_impl(url, destfile),
        condition = identity
      )
  
      if (identical(status, 0L)) {
        message("OK")
        return(destfile)
      }
  
    }
  
    message("FAILED")
    return(FALSE)
  
  }
  
  renv_bootstrap_download_github <- function(version) {
  
    enabled <- Sys.getenv("RENV_BOOTSTRAP_FROM_GITHUB", unset = "TRUE")
    if (!identical(enabled, "TRUE"))
      return(FALSE)
  
    # prepare download options
    pat <- Sys.getenv("GITHUB_PAT")
    if (nzchar(Sys.which("curl")) && nzchar(pat)) {
      fmt <- "--location --fail --header \"Authorization: token %s\""
      extra <- sprintf(fmt, pat)
      saved <- options("download.file.method", "download.file.extra")
      options(download.file.method = "curl", download.file.extra = extra)
      on.exit(do.call(base::options, saved), add = TRUE)
    } else if (nzchar(Sys.which("wget")) && nzchar(pat)) {
      fmt <- "--header=\"Authorization: token %s\""
      extra <- sprintf(fmt, pat)
      saved <- options("download.file.method", "download.file.extra")
      options(download.file.method = "wget", download.file.extra = extra)
      on.exit(do.call(base::options, saved), add = TRUE)
    }
  
    message("* Downloading renv ", version, " from GitHub ... ", appendLF = FALSE)
  
    url <- file.path("https://api.github.com/repos/rstudio/renv/tarball", version)
    name <- sprintf("renv_%s.tar.gz", version)
    destfile <- file.path(tempdir(), name)
  
    status <- tryCatch(
      renv_bootstrap_download_impl(url, destfile),
      condition = identity
    )
  
    if (!identical(status, 0L)) {
      message("FAILED")
      return(FALSE)
    }
  
    message("Done!")
    return(destfile)
  
  }
  
  renv_bootstrap_install <- function(version, tarball, library) {
  
    # attempt to install it into project library
    message("* Installing renv ", version, " ... ", appendLF = FALSE)
    dir.create(library, showWarnings = FALSE, recursive = TRUE)
  
    # invoke using system2 so we can capture and report output
    bin <- R.home("bin")
    exe <- if (Sys.info()[["sysname"]] == "Windows") "R.exe" else "R"
    r <- file.path(bin, exe)
    args <- c("--vanilla", "CMD", "INSTALL", "-l", shQuote(library), shQuote(tarball))
    output <- system2(r, args, stdout = TRUE, stderr = TRUE)
    message("Done!")
  
    # check for successful install
    status <- attr(output, "status")
    if (is.numeric(status) && !identical(status, 0L)) {
      header <- "Error installing renv:"
      lines <- paste(rep.int("=", nchar(header)), collapse = "")
      text <- c(header, lines, output)
      writeLines(text, con = stderr())
    }
  
    status
  
  }
  
  renv_bootstrap_prefix <- function() {
  
    # construct version prefix
    version <- paste(R.version$major, R.version$minor, sep = ".")
    prefix <- paste("R", numeric_version(version)[1, 1:2], sep = "-")
  
    # include SVN revision for development versions of R
    # (to avoid sharing platform-specific artefacts with released versions of R)
    devel <-
      identical(R.version[["status"]],   "Under development (unstable)") ||
      identical(R.version[["nickname"]], "Unsuffered Consequences")
  
    if (devel)
      prefix <- paste(prefix, R.version[["svn rev"]], sep = "-r")
  
    # build list of path components
    components <- c(prefix, R.version$platform)
  
    # include prefix if provided by user
    prefix <- Sys.getenv("RENV_PATHS_PREFIX")
    if (nzchar(prefix))
      components <- c(prefix, components)
  
    # build prefix
    paste(components, collapse = "/")
  
  }
  
  renv_bootstrap_library_root <- function(project) {
  
    path <- Sys.getenv("RENV_PATHS_LIBRARY", unset = NA)
    if (!is.na(path))
      return(path)
  
    path <- Sys.getenv("RENV_PATHS_LIBRARY_ROOT", unset = NA)
    if (!is.na(path)) {
      id <- substring(renv_bootstrap_hash_text(project), 1L, 8L)
      name <- paste(basename(project), id, sep = "-")
      return(file.path(path, name))
    }
  
    file.path(project, "renv/library")
  
  }
  
  renv_bootstrap_validate_version <- function(version) {
  
    loadedversion <- utils::packageDescription("renv", fields = "Version")
    if (version == loadedversion)
      return(TRUE)
  
    # assume four-component versions are from GitHub; three-component
    # versions are from CRAN
    components <- strsplit(loadedversion, "[.-]")[[1]]
    remote <- if (length(components) == 4L)
      paste("rstudio/renv", loadedversion, sep = "@")
    else
      paste("renv", loadedversion, sep = "@")
  
    fmt <- paste(
      "renv %1$s was loaded from project library, but this project is configured to use renv %2$s.",
      "Use `renv::record(\"%3$s\")` to record renv %1$s in the lockfile.",
      "Use `renv::restore(packages = \"renv\")` to install renv %2$s into the project library.",
      sep = "\n"
    )
  
    msg <- sprintf(fmt, loadedversion, version, remote)
    warning(msg, call. = FALSE)
  
    FALSE
  
  }
  
  renv_bootstrap_hash_text <- function(text) {
  
    hashfile <- tempfile("renv-hash-")
    on.exit(unlink(hashfile), add = TRUE)
  
    writeLines(text, con = hashfile)
    tools::md5sum(hashfile)
  
  }
  
  renv_bootstrap_load <- function(project, libpath, version) {
  
    # try to load renv from the project library
    if (!requireNamespace("renv", lib.loc = libpath, quietly = TRUE))
      return(FALSE)
  
    # warn if the version of renv loaded does not match
    renv_bootstrap_validate_version(version)
  
    # load the project
    renv::load(project)
  
    TRUE
  
  }

  # construct path to library root
  root <- renv_bootstrap_library_root(project)

  # construct library prefix for platform
  prefix <- renv_bootstrap_prefix()

  # construct full libpath
  libpath <- file.path(root, prefix)

  # attempt to load
  if (renv_bootstrap_load(project, libpath, version))
    return(TRUE)

  # load failed; inform user we're about to bootstrap
  prefix <- paste("# Bootstrapping renv", version)
  postfix <- paste(rep.int("-", 77L - nchar(prefix)), collapse = "")
  header <- paste(prefix, postfix)
  message(header)

  # perform bootstrap
  bootstrap(version, libpath)

  # exit early if we're just testing bootstrap
  if (!is.na(Sys.getenv("RENV_BOOTSTRAP_INSTALL_ONLY", unset = NA)))
    return(TRUE)

  # try again to load
  if (requireNamespace("renv", lib.loc = libpath, quietly = TRUE)) {
    message("* Successfully installed and loaded renv ", version, ".")
    return(renv::load())
  }

  # failed to download or load renv; warn the user
  msg <- c(
    "Failed to find an renv installation: the project will not be loaded.",
    "Use `renv::activate()` to re-initialize the project."
  )

  warning(paste(msg, collapse = "\n"), call. = FALSE)

})


data-raw/hpmvec-processing.R
devtools::load_all()

# setup parallel processing
doParallel::registerDoParallel(parallel::detectCores())
BiocParallel::register(BiocParallel::DoparParam(), default = TRUE)

# raw data from hPMVEC exposed to 21% or 0.5% oxygen for 24 h
# collected on Q Exactive with ZIC-pHILIC separation
# top 10 data-dependent scanning in negative polarity mode
# three 21% and three 0.5% oxygen samples selected as sample dataset

# get full path of data files
files <-
  list.files(path = system.file("extdata", package = "IPO2"),
             pattern = "_1_",
             full.names = TRUE,
             recursive = TRUE)

# create a phenodata data.frame
pd <-
  data.frame(id = stringr::str_extract(basename(files), pattern = "^\\d{2}"),
             group = stringr::str_extract(basename(files), pattern = "(norm|hyp)"),
             stringsAsFactors = FALSE)

# read raw data
raw_data <-
  MSnbase::readMSData(files = files,
                      pdata = new("NAnnotatedDataFrame", pd),
                      msLevel. = 1L,
                      mode = "onDisk")

# perform peak picking with centWave algorithm
cwp <- xcms::CentWaveParam(peakwidth = c(20, 80), noise = 5000)
hpmvec <- xcms::findChromPeaks(raw_data, param = cwp)

usethis::use_data(hpmvec, overwrite = TRUE)


R/optimize_centwave.R
#' Find centWave parameters
#'
#' `optimize_centwave` will identify parameters for `xcms::findChromPeaks` by
#' optimizing the identification of isotope pairs. Note that currently the
#' isotope identification algorithm relies on high resolution m/*z* measurements.
#'
#' @param raw_data MSnExp object containing the raw data
#' @param parameter_list List of parameters to set or to optimize
#' @param out_dir Path for output directory. Defaults to working directory.
#'
#' @return A list of length 2 containing:
#' \describe{
#'   \item{history}{a data.table object recording the optimization history}
#'   \item{best_cwp}{a CentWaveParam object of the optimal parameters}
#' }
#' @export
#'
#' @examples
#' \dontrun{
#' files <- list.files(system.file("extdata", package = "IPO2"), full.names = TRUE)
#' raw_data <-
#'   MSnbase::readMSData(files = files[[1]], mode = "onDisk") %>%
#'   MSnbase::filterRt(c(300, 900))
#' cwp_params <- optimize_centwave(raw_data = raw_data)
#' }
#'
optimize_centwave <- function(
  raw_data = NULL,
  parameter_list = suggest_centwave_params(),
  out_dir = NULL
) {

  log_file <- redo_list <- NULL

  # prepare output
  prepare_out_dir(out_dir = out_dir, fun_name = "centwave")
  sink(log_file, append = TRUE, type = "output", split = TRUE)
  on.exit(sink(), add = TRUE, after = TRUE)

  # check parameters
  check_centwave_params(parameter_list)

  # set up iteration loop
  history <- list()
  iteration <- 1
  while(iteration <= 50) {

    cat(
      format(Sys.time(), "%Y-%m-%d %H:%M:%S"),
      ": Starting iteration ",
      iteration,
      "\n"
    )

    # parse parameters
    parameters <- parse_parameters(parameter_list, "centwave")

    # design experiments
    design <- design_experiments(parameters)

    # generate centWave parameters
    cwp <-
      purrr::pmap(
        cbind(design, t(parameters$constant), t(parameters$lists)),
        make_cwp
      )

    # run xcms for each experiment
    score <- rbindlist(
      retry_parallel(
        rlang::expr(
          BiocParallel::bplapply(
            cwp,
            function(x) {
              score_peaks(
                xcms::findChromPeaks(
                  raw_data,
                  param = x,
                  BPPARAM = BiocParallel::SerialParam()
                )
              )
            },
            BPREDO = redo_list
          )
        )
      )
    )

    # calculate model
    model <- create_model(design, score)

    # find best parameters
    maximum <- get_maximum(design, model)

    # make plots
    plot_name <-
      paste0(dir, "/contour_centwave_", sprintf("%02d", iteration), ".png")
    plot_contours(design, model, maximum, plot_name)

    # score best parameters
    max_cwp <- purrr::pmap(
      data.table(t(maximum), t(parameters$constant), t(parameters$lists)),
      make_cwp
    )
    max_xcmsnexp <- xcms::findChromPeaks(raw_data, param = max_cwp[[1]], BPPARAM = BiocParallel::SerialParam())
    max_score <- score_peaks(max_xcmsnexp)

    # assign to history
    history[[iteration]] <-
      c(
        unlist(parameters$to_optimize),
        cwp = list(max_cwp),
        maximum,
        max_score
      )

    hx <- rbindlist(history)[, !c("cwp")]
    end_cols <- c("total", "isotopes", "indeterminate", "score")
    setcolorder(hx, order(setdiff(names(hx), end_cols)))

    fwrite(hx, paste0(dir, "/history.csv"), row.names = TRUE)

    cat("\n")

    # check for improvement
    better <- hx[[iteration, "score"]] == max(hx[["score"]])

    # adjust intervals
    if (better) {
      new_parameters <- pick_parameters(parameters$to_optimize, maximum)
      parameter_list[names(new_parameters)] <- new_parameters
    } else {
      break
    }

    # increase counter
    iteration <- iteration + 1

  }

  # output results
  history <- rbindlist(history)
  setcolorder(history, order(setdiff(names(history), c(end_cols, cwp))))
  best_cwp <- history[score == max(score), cwp][[1]]
  list(history = history, best_cwp = best_cwp)

}


get_centwave_defaults <- function() {
  list(
    ppm                = 25,
    min_peakwidth      = 20,
    max_peakwidth      = 50,
    snthresh           = 10,
    prefilter_k        = 3,
    prefilter_int      = 100,
    mzdiff             = -0.001,
    noise              = 0,
    mzCenterFun        = "wMean",
    integrate          = 1L,
    fitgauss           = FALSE,
    verboseColumns     = FALSE,
    roiList            = list(),
    firstBaselineCheck = TRUE,
    roiScales          = numeric()
  )
}


suggest_centwave_params <- function() {
  list(
    ppm                = c(17, 32),
    min_peakwidth      = c(12, 28),
    max_peakwidth      = c(35, 65),
    snthresh           = 100,  # originally 10, increase to increase speed
    prefilter_k        = 3,
    prefilter_int      = 10000,  # originally 100
    mzdiff             = c(-0.001, 0.01),
    noise              = 0  # originally 0, increase to increase speed
  )
}


param_types_centwave <- function() {

  assign(
    "quant",
    c(
      "ppm",
      "min_peakwidth",
      "max_peakwidth",
      "snthresh",
      "prefilter_k",
      "prefilter_int",
      "mzdiff",
      "noise"
    ),
    envir = parent.frame()
  )

  assign(
    "qual",
    c(
      "mzCenterFun",
      "integrate",
      "fitgauss",
      "verboseColumns",
      "firstBaselineCheck"
    ),
    envir = parent.frame()
  )

  assign(
    "lists",
    c(
      "roiList",
      "roiScales"
    ),
    envir = parent.frame()
  )

}


check_centwave_params <- function(parameter_list) {

  quant <- qual <- lists <- NULL

  param_types_centwave()

  # check parameter names
  nms <- names(parameter_list)
  bad_names <- nms[nms %nin% c(quant, qual, lists)]
  if (length(bad_names) != 0) {
    stop("These parameters are not used by centWave: ",
         paste(bad_names, collapse = ", "))
  }

  # check if too many parameters provided
  lengths <- lapply(parameter_list[nms %in% quant], length)
  bad_lengths <- names(lengths[lengths > 2])
  if (any(lengths > 2)) {
    stop("Too many parameters specified for: ",
         paste(bad_lengths, collapse = ", "))
  }

  # check for parameters to optimize
  if (!(any(lengths == 2))) {
    stop("No parameters specified for optimization")
  }

  # check ranges
  for (nm in nms){
    if (nm %in% c("ppm")) {
      if (min(parameter_list[[nm]]) <= 0) {
        stop(paste("The parameter", nm, "must be greater than 0"))
      }
    } else if (nm %in% c("min_peakwidth", "max_peakwidth")) {
      if (min(parameter_list[[nm]]) <= 0) {
        stop(paste("The parameter", nm, "must be greater than 0"))
      }
    } else if (nm %in% c("snthresh", "prefilter_k", "prefilter_int", "noise")) {
      if (min(parameter_list[[nm]]) < 0) {
        stop(paste("The parameter", nm, "must not be negative"))
      }
    }
  }
}


make_cwp <- function(
  min_peakwidth = 20,
  max_peakwidth = 50,
  prefilter_k = 3,
  prefilter_int = 100,
  ...
) {
  xcms::CentWaveParam(
    peakwidth = c(min_peakwidth, max_peakwidth),
    prefilter = c(prefilter_k, prefilter_int),
    ...)
}
